chatsky/utils/context_dict/ctx_dict.py

```diff
@@ -0,0 +1,203 @@
+        hashes = {k: hash(v) for k, v in items.items()}
+        instance = cls.model_validate(items)
+        instance._storage = storage
+        instance._ctx_id = id
+        instance._field_name = field
+        instance._field_constructor = constructor
+        instance._keys = keys
+        instance._hashes = hashes
+        return instance
+
+    async def _load_items(self, keys: List[K]) -> Dict[K, V]:
+        items = await self._storage.load_field_items(self._ctx_id, self._field_name, keys)
+        for key, item in zip(keys, items):
+            self._items[key] = self._field_constructor(item)
+            self._hashes[key] = hash(item)
+
+    async def __getitem__(self, key: Union[K, slice]) -> V:
+        if self._storage is not None and self._storage.rewrite_existing:
+            if isinstance(key, slice):
+                await self._load_items([k for k in range(len(self._keys))[key] if k not in self._items.keys()])
+            elif key not in self._items.keys():
+                await self._load_items([key])
+        if isinstance(key, slice):
+            return {k: await self._items[k] for k in range(len(self._items.keys()))[key]}
+        else:
+            return self._items[key]
+
+    def __setitem__(self, key: Union[K, slice], value: Union[V, Sequence[V]]) -> None:
+        if isinstance(key, slice) and isinstance(value, Sequence):
+            if len(key) != len(value):
+                raise ValueError("Slices must have the same length!")
+            for k, v in zip(range(len(self._keys))[key], value):
+                self[k] = v
+        elif not isinstance(key, slice) and not isinstance(value, Sequence):
+            self._keys += [key]
+            if key not in self._items.keys():
+                self._added += [key]
+            if key in self._removed:
+                self._removed.remove(key)
+            self._items[key] = value
+        else:
+            raise ValueError("Slice key must have sequence value!")
+
+    def __delitem__(self, key: Union[K, slice]) -> None:
+        if isinstance(key, slice):
+            for k in range(len(self._keys))[key]:
+                del self[k]
+        else:
+            self._removed += [key]
+            if key in self._items.keys():
+                self._keys.remove(key)
+            if key in self._added:
+                self._added.remove(key)
+            del self._items[key]
+
+    def __iter__(self) -> Sequence[K]:
+        return iter(self._keys if self._storage is not None else self._items.keys())
+    
+    def __len__(self) -> int:
+        return len(self._keys if self._storage is not None else self._items.keys())
+
+    async def get(self, key: K, default: V = _marker) -> V:
+        try:
+            return await self[key]
+        except KeyError:
+            if default is self._marker:
+                raise
```

Comment:
Is marker used to forbid not passing `default`?

Reply:
Why not make default `None`?

Reply:
But the standard `dict` returns `None` by default if key doesn't exist.

If I want an exception to be thrown on missing key I use the `__getitem__` method instead.